File: python/283_MoveZeroes.py
class Solution:
    def moveZeroes(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        count, idx = 0, 0
        while idx < len(nums):
            if nums[idx] == 0:
                del nums[idx]
                count += 1
            else:
                idx += 1
        for i in range(count):
            nums.append(0)
        return nums

# 30 day challenge solution
class Solution:
    def moveZeroes(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        i, count = 0, 0
        while i < len(nums):
            if nums[i] == 0:
                count += 1
                del nums[i]
            else:
                i += 1
        
        for i in range(count):
            nums.append(0)

File: java/283_MoveZeroes.java
class Solution {
    public void moveZeroes(int[] nums) {
        int nonZero = 1;
        for (int i = 0; i < nums.length; ++i) {
            if (nums[i] == 0) {
                while (nonZero < nums.length && nums[nonZero] == 0) ++nonZero;
                if (nonZero < nums.length && nums[nonZero] != 0) {
                    nums[i] = nums[nonZero];
                    nums[nonZero++] = 0;
                }
            } else {
                ++nonZero;
            }
        }
    }
}
